File: backend/src/tasks/tasks.service.ts
import TasksModel from "./tasks.model";
import { logger } from "../config/logger";

export const createTaskService = async (
  userId: string,
  taskData: { title: string; description: string; priority: string; status: string; dueDate: string; category: string; progress?: number }
) => {
  try {
    logger.info("Creating task for user => " + userId);
    const task = await TasksModel.create({
      userId,
      ...taskData,
    });
    logger.info("Task created successfully => " + task._id);
    return task;
  } catch (error) {
    logger.error("Task creation failed => " + error);
    throw Error("Task creation failed in service");
  }
};

export const getAllTasksService = async (userId: string) => {
  try {
    logger.info("Fetching all tasks for user => " + userId);
    // Filter deletedAt: null to exclude soft-deleted tasks
    const tasks = await TasksModel.find({ userId, deletedAt: null }).sort({ createdAt: -1 });
    logger.info("Tasks fetched successfully, total => " + tasks.length);
    return tasks;
  } catch (error) {
    logger.error("Failed to fetch tasks => " + error);
    throw Error("Failed to fetch tasks in service");
  }
};

export const getTaskByIdService = async (taskId: string, userId: string) => {
  try {
    logger.info("Fetching task => " + taskId);
    // Ensure task belongs to user and is not deleted
    const task = await TasksModel.findOne({ _id: taskId, userId, deletedAt: null });
    if (!task) {
      logger.error("Task not found => " + taskId);
      return null;
    }
    logger.info("Task fetched successfully");
    return task;
  } catch (error) {
    logger.error("Failed to fetch task => " + error);
    throw Error("Failed to fetch task in service");
  }
};

export const updateTaskService = async (
  taskId: string,
  userId: string,
  updateData: { title?: string; description?: string; priority?: string; status?: string; dueDate?: string; category?: string; progress?: number }
) => {
  try {
    logger.info("Updating task => " + taskId + " for user => " + userId);
    // Using findOneAndUpdate with new: true to return updated document
    const task = await TasksModel.findOneAndUpdate(
      { _id: taskId, userId, deletedAt: null },
      updateData,
      { new: true }
    );
    if (!task) {
      logger.error("Task not found for update => " + taskId);
      return null;
    }
    logger.info("Task updated successfully => " + taskId);
    return task;
  } catch (error) {
    logger.error("Failed to update task => " + error);
    throw Error("Failed to update task in service");
  }
};

export const deleteTaskService = async (taskId: string, userId: string) => {
  try {
    logger.info("Deleting task => " + taskId + " for user => " + userId);
    // Soft delete: setting deletedAt instead of removing from DB
    const task = await TasksModel.findOneAndUpdate(
      { _id: taskId, userId, deletedAt: null },
      { deletedAt: new Date() },
      { new: true }
    );
    if (!task) {
      logger.error("Task not found for deletion => " + taskId);
      return null;
    }
    logger.info("Task deleted successfully => " + taskId);
    return task;
  } catch (error) {
    logger.error("Failed to delete task => " + error);
    throw Error("Failed to delete task in service");
  }
};

File: frontend/dev_vault/lib/data/services/tasks_service.dart
import 'package:dio/dio.dart';
import '../constant_urls.dart';
import '../models/tasks_model.dart';
import 'local_storage_service.dart';

class TasksService {
  static final Dio _dio = Dio();

  /// Get authorization headers with stored token
  static Map<String, dynamic> _getAuthHeaders() {
    final token = LocalStorageService.getToken();
    return {'Authorization': token != null ? 'Bearer $token' : ''};
  }

  /// Get all tasks
  static Future<List<TasksModel>> getAllTasks() async {
    try {
      print('📡 Fetching all tasks from: $getAllTasksUrl');
      print('🔑 Using stored token for authentication');

      final response = await _dio.get(
        getAllTasksUrl,
        options: Options(headers: _getAuthHeaders()),
      );

      if (response.statusCode == 200 || response.statusCode == 201) {
        final data = response.data;

        if (data['success'] == true && data['data'] != null) {
          final tasksList = (data['data'] as List)
              .map((task) => TasksModel.fromJson(task))
              .toList();
          print('✅ Tasks fetched successfully');
          print('   Total tasks: ${tasksList.length}');
          return tasksList;
        } else {
          throw Exception(data['message'] ?? 'Failed to fetch tasks');
        }
      } else {
        throw Exception('Failed to fetch tasks with status ${response.statusCode}');
      }
    } on DioException catch (e) {
      print('❌ Dio Error: ${e.message}');
      if (e.response?.data != null) {
        throw Exception(e.response?.data['message'] ?? e.message);
      }
      throw Exception('Error fetching tasks: ${e.message}');
    } catch (e) {
      print('❌ Exception: $e');
      throw Exception(e.toString());
    }
  }

  /// Get task by ID
  static Future<TasksModel> getTaskById(String id) async {
    try {
      print('📡 Fetching task with ID: $id');

      final response = await _dio.get(
        '$getTaskByIdUrl$id',
        options: Options(headers: _getAuthHeaders()),
      );

      if (response.statusCode == 200 || response.statusCode == 201) {
        final data = response.data;

        if (data['success'] == true && data['data'] != null) {
          final task = TasksModel.fromJson(data['data']);
          print('✅ Task fetched successfully: ${task.title}');
          return task;
        } else {
          throw Exception(data['message'] ?? 'Failed to fetch task');
        }
      } else {
        throw Exception('Failed to fetch task with status ${response.statusCode}');
      }
    } on DioException catch (e) {
      print('❌ Dio Error: ${e.message}');
      if (e.response?.data != null) {
        throw Exception(e.response?.data['message'] ?? e.message);
      }
      throw Exception('Error fetching task: ${e.message}');
    } catch (e) {
      print('❌ Exception: $e');
      throw Exception(e.toString());
    }
  }

  /// Create new task
  static Future<TasksModel> createTask({
    required String title,
    required String description,
    required String priority,
    required String status,
    required String dueDate,
    required String category,
    int progress = 0,
  }) async {
    try {
      print('📡 Creating new task');
      print('   Title: $title');
      print('   Priority: $priority');
      print('   Category: $category');

      final response = await _dio.post(
        createTaskUrl,
        data: {
          'title': title,
          'description': description,
          'priority': priority,
          'status': status,
          'dueDate': dueDate,
          'category': category,
          'progress': progress,
        },
        options: Options(headers: _getAuthHeaders()),
      );

      if (response.statusCode == 200 || response.statusCode == 201) {
        final data = response.data;

        if (data['success'] == true && data['data'] != null) {
          final task = TasksModel.fromJson(data['data']);
          print('✅ Task created successfully: ${task.title}');
          return task;
        } else {
          throw Exception(data['message'] ?? 'Failed to create task');
        }
      } else {
        throw Exception('Failed to create task with status ${response.statusCode}');
      }
    } on DioException catch (e) {
      print('❌ Dio Error: ${e.message}');
      if (e.response?.data != null) {
        throw Exception(e.response?.data['message'] ?? e.message);
      }
      throw Exception('Error creating task: ${e.message}');
    } catch (e) {
      print('❌ Exception: $e');
      throw Exception(e.toString());
    }
  }

  /// Update task
  static Future<TasksModel> updateTask({
    required String id,
    required String title,
    required String description,
    required String priority,
    required String status,
    required String dueDate,
    required String category,
    int progress = 0,
  }) async {
    try {
      print('📡 Updating task with ID: $id');
      print('   Title: $title');
      print('   Status: $status');

      final response = await _dio.put(
        '$updateTaskUrl$id',
        data: {
          'title': title,
          'description': description,
          'priority': priority,
          'status': status,
          'dueDate': dueDate,
          'category': category,
          'progress': progress,
        },
        options: Options(headers: _getAuthHeaders()),
      );

      if (response.statusCode == 200 || response.statusCode == 201) {
        final data = response.data;

        if (data['success'] == true && data['data'] != null) {
          final task = TasksModel.fromJson(data['data']);
          print('✅ Task updated successfully: ${task.title}');
          return task;
        } else {
          throw Exception(data['message'] ?? 'Failed to update task');
        }
      } else {
        throw Exception('Failed to update task with status ${response.statusCode}');
      }
    } on DioException catch (e) {
      print('❌ Dio Error: ${e.message}');
      if (e.response?.data != null) {
        throw Exception(e.response?.data['message'] ?? e.message);
      }
      throw Exception('Error updating task: ${e.message}');
    } catch (e) {
      print('❌ Exception: $e');
      throw Exception(e.toString());
    }
  }

  /// Delete task
  static Future<void> deleteTask(String id) async {
    try {
      print('📡 Deleting task with ID: $id');

      final response = await _dio.delete(
        '$deleteTaskUrl$id',
        options: Options(headers: _getAuthHeaders()),
      );

      if (response.statusCode == 200 || response.statusCode == 201) {
        final data = response.data;

        if (data['success'] == true) {
          print('✅ Task deleted successfully');
        } else {
          throw Exception(data['message'] ?? 'Failed to delete task');
        }
      } else {
        throw Exception('Failed to delete task with status ${response.statusCode}');
      }
    } on DioException catch (e) {
      print('❌ Dio Error: ${e.message}');
      if (e.response?.data != null) {
        throw Exception(e.response?.data['message'] ?? e.message);
      }
      throw Exception('Error deleting task: ${e.message}');
    } catch (e) {
      print('❌ Exception: $e');
      throw Exception(e.toString());
    }
  }
}
